### ISSUE TITLE
Implement Mirroring support

### ISSUE THREAD

**Issue by wzin** [EndPointCorp/lg_ros_nodes#100]
tldr - mirroring means: setting up tuple of `gstreamer capture`, `gstreamer playback` (or receiver), and `mirrored app` (e.g.chrome) that `gstreamer capture` captures and `gstreamer playback` plays back. Touchscreen touch is typically routed from the touchscreen to `mirrored app`. 

**Requirements**:
- make `mirror` ROS node that will run on every host
- make `mirror` react on USCS with `activity` of `mirror`
- if asset equals `viewport://<viewport_name>` then start capturing this viewport e.g.
```json
{
      "activity": "mirror",
      "assets": [
        "viewport://center_portrait_three"
      ],
      "activity_config": {
       },
      "height": 1920,
      "presentation_viewport": "center_portrait_three",
      "width": 3240,
      "x_coord": 0,
      "y_coord": 0
    }
```
 - if asset equals `mirror://42-a/<viewport_name>` then start play back from this viewport on viewport specified in the `window` - example message:
```json
 {
      "activity": "mirror",
      "assets": [
        "mirror://42-a/center_portrait_three"
      ],
      "activity_config": {
       },
      "height": 1080,
      "presentation_viewport": "center_portrait_three",
      "width": 1824,
      "x_coord": 76,
      "y_coord": 0
    }
```
- maintain pool of gstreamers just like `AdhocBrowserPool` does it
- make sure that aspect ratio of the playback window and capture window are correctly matched using matrix transformations (xinput)
- OPTIONAL BONUS POINTS: make support for running whatever number of mirroring tuples (`gst capture` + `gst playback` + `mirrored app`) - e.g. to be able to mirror whole LG on a touchscreen - this means that gstreamers need to be unique and use unique ports and slugs

**Development notes**:
- eample message https://github.com/EndPointCorp/lg_ros_nodes/blob/master/interactivespaces_msgs/example_messages/wall_kiosk.json
- running matrix transformation commands can be done via xinput:
 - xinput -> 3240x1080 -> 100x200+50+50
https://github.com/EndPointCorp/lg-live-image/blob/portal/config/includes.chroot/etc/X11/Xsession.d/45x11-custom_lg-xinput#L34-L43
 - 1080x1080 -> 3240x1080 - evdev teleport receiver - call xinput to make matrix transformations

**Comment by mvollrath** [comment]
[Much progress was made](https://github.com/EndPointCorp/lg_ros_nodes/commit/5d8992fb9b15f85ad5a930d5ad3bff7158ba2269) on the capture side.

I've simplified the USCS integration a bit.  A `mirror` activity represents both the capture and playback.  For example, from the tests:

```
{
  "description": "bogus",
  "duration": 0,
  "name": "test whatever",
  "resource_uri": "bogus",
  "slug": "test message",
  "windows": [
    {
      "activity": "mirror",
      "activity_config": {
      },
      "assets": [
        "viewport://capture_vport"
      ],
      "presentation_viewport": "playback_vport",
      "width": 1920,
      "height": 1080,
      "x_coord": 0,
      "y_coord": 0
    }
  ]
}
```

Due to GStreamer CLI limitations, I'm moving forward with *a maximum of one mirror activity in any given scene.*  Because GStreamer's `xvimagesink` always sets the same window name, we can't hook it up to window management unless we make a custom binary with the GStreamer C API and call [g_set_application_name()](https://developer.gnome.org/glib/stable/glib-Miscellaneous-Utility-Functions.html#g-set-application-name) before creating the sink.  Without a good use case for it, I can't justify our maintaining such a binary.

**Comment by mvollrath** [comment]
I was wrong!  By using the [taginject](https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-good-plugins/html/gst-plugins-good-plugins-taginject.html) plugin, we can set the media title which is respected by `xvimagesink`.

For example,

`gst-launch-1.0 videotestsrc is-live=true ! taginject tags="title=foo" ! xvimagesink`

This results in a window name of `"foo : gst-launch-1.0"`

+1 for [reading the code](https://github.com/rawoul/gst-plugins-base/blob/e05abf0ef19676b0be3a0e43c8ca1f75910d2093/sys/xvimage/xvimagesink.c#L1918-L1932) :godmode:

**Comment by mvollrath** [comment]
Left to do:

* Upscaling video on destination window (if necessary)
* Explanation/examples of Scene parsing in README
* CMS integration
* Testing

**Comment by mvollrath** [comment]
Mirroring has been merged.  https://github.com/EndPointCorp/lg_ros_nodes/pull/271

**Comment by mvollrath** [comment]
Lumping touch teleport into this.  The plan is to manage per-viewport receivers with a touch router listening for scene changes.  Something like:

### Sender (one per graph)

#### Params
* device_path

#### Publications
* /lg_mirror/events (`lg_mirror/EvdevEvents`)

#### Services
* /lg_mirror/device_info (`lg_mirror/EvdevDeviceInfo`)

### Receiver (one per viewport)

#### Params
* viewport

#### Subscriptions
* /lg_mirror/events (`lg_mirror/EvdevEvents`)
* /lg_mirror/touch_viewport (`std_msgs/String`)

### Router (one per graph)

#### Params
* default_viewport (defaults to "touchscreen")

#### Subscriptions
* /director/scene (USCS)

#### Publications
* /lg_mirror/touch_viewport (`std_msgs/String`)

#### Logic
Upon a new scene, if the scene contains a "mirror" window and the activity_config includes `route_touch: true`, route touches to the source viewport.  Otherwise, route touches to the default viewport.

**Comment by wzin** [comment]
Some findings: https://www.flowdock.com/app/endpoint/r-d/threads/65wJaW7LGjpoJSOab6344TIvyK4

**Comment by mvollrath** [comment]
The last kludge on this feature is that fullscreen playback windows are sometimes placed on the wrong screen.

**Comment by mvollrath** [comment]
There appears to be a race between xvimagesink and awesome causing the intermittent failure to place windows properly.

When analyzing this debug output:

```
$ gst-launch-1.0 --gst-debug=xvimagesink:5 udpsrc address=224.42.42.1 port=4953 ! capsfilter caps=application/x-rtp,payload=96 ! rtpjpegdepay 
! jpegdec ! taginject tags="title=foo" ! xvimagesink display=:0 sync=false                                                                                             
Setting pipeline to PAUSED ...
0:00:00.016777917 25582      0x1123350 DEBUG            xvimagesink xvimagesink.c:1751:gst_xvimagesink_open:<xvimagesink0> set calculated PAR on object's PAR
0:00:00.016807471 25582      0x1123350 DEBUG            xvimagesink xvimagesink.c:606:gst_xvimagesink_manage_event_thread:<xvimagesink0> run xevent thread, expose 1, events 1
Pipeline is live and does not need PREROLL ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
0:00:00.030711897 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:678:gst_xvimagesink_setcaps:<xvimagesink0> In setcaps. Possible caps video/x-raw, format=(string)YUY2, width=(int)[ 1, 32768 ], height=(int)[ 1, 32768 ], framerate=(fraction)[ 0/1, 2147483647/1 ]; video/x-raw, format=(string)YV12, width=(int)[ 1, 32768 ], height=(int)[ 1, 32768 ], framerate=(fraction)[ 0/1, 2147483647/1 ]; video/x-raw, format=(string)UYVY, width=(int)[ 1, 32768 ], height=(int)[ 1, 32768 ], framerate=(fraction)[ 0/1, 2147483647/1 ]; video/x-raw, format=(string)I420, width=(int)[ 1, 32768 ], height=(int)[ 1, 32768 ], framerate=(fraction)[ 0/1, 2147483647/1 ], setting caps video/x-raw, format=(string)I420, width=(int)456, height=(int)800, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive, colorimetry=(string)1:4:0:0, framerate=(fraction)0/1
0:00:00.030834557 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:723:gst_xvimagesink_setcaps:<xvimagesink0> video width/height: 456x800, calculated display ratio: 57/100
0:00:00.030864238 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:732:gst_xvimagesink_setcaps:<xvimagesink0> keeping video height
0:00:00.030882835 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:748:gst_xvimagesink_setcaps:<xvimagesink0> scaling to 456x800
0:00:00.031257857 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:1062:gst_xvimagesink_propose_allocation:<xvimagesink0> check existing pool caps
0:00:00.035482698 25582      0x112f190 DEBUG            xvimagesink xvimagesink.c:1024:gst_xvimagesink_event:<xvimagesink0> got tags, title='foo'
0:00:00.067707371 25582      0x112f230 DEBUG            xvimagesink xvimagesink.c:1257:gst_xvimagesink_expose: doing expose
```

We see that `gst_xvimagesink_setcaps` happens before the `gst_xvimagesink_event`.  The [window is created in `gst_xvimagesink_setcaps()`](https://github.com/rawoul/gst-plugins-base/blob/e05abf0ef19676b0be3a0e43c8ca1f75910d2093/sys/xvimage/xvimagesink.c#L1645-L1648) with its [name set to the default](https://github.com/rawoul/gst-plugins-base/blob/e05abf0ef19676b0be3a0e43c8ca1f75910d2093/sys/xvimage/xvimagesink.c#L496-L497) (which is the application name, "gst-launch-1.0").

In the time between the window being created and the title being set according to the tag event, awesome is handling the new window event and trying to match it against its rules table.  When xvimagesink wins, the window is in the right place.  When awesome wins, the window is unmanaged.

I tried using the [awesome signals API](https://awesome.naquadah.org/wiki/Signals#client) to consider its rules whenever a client changes its `property:name`.  This snippet in rc.lua results in consistently managed windows:

```
client.connect_signal("manage", function (c, startup)
  c:connect_signal("property::name", awful.rules.apply)
end)
```

But it still doesn't put the windows in the right place.

Since the use cases so far only call for one GStreamer instance at a time anywhere, I'm removing support for multiple playback windows for now and going back to matching the "gst-launch-1.0" window name.
